src/constructors/regular.c

```diff
@@ -408,6 +407,103 @@ igraph_error_t igraph_tree(igraph_t *graph, igraph_integer_t n, igraph_integer_t
     return IGRAPH_SUCCESS;
 }
 
+/**
+ * \ingroup generators
+ * \function igraph_symmetric_tree
+ * \brief Creates a symmetric tree which has the same number of branches at each level.
+ *
+ * Cannot create null graph with this function because the root vertex always gets created 
+ * since an empty vector creates a singleton graph not a null graph.
+ *
+ * \param graph Pointer to an uninitialized graph object.
+ * \param vector Vector detailing the number of branches at each level.
+ * \param type Constant, gives whether to create a directed tree, and
+ *        if this is the case, also its orientation. Possible values:
+ *        \clist
+ *        \cli IGRAPH_TREE_OUT
+ *          directed tree, the edges point
+ *          from the parents to their children,
+ *        \cli IGRAPH_TREE_IN
+ *          directed tree, the edges point from
+ *          the children to their parents.
+ *        \cli IGRAPH_TREE_UNDIRECTED
+ *          undirected tree.
+ *        \endclist
+ * \return Error code:
+ *         \c IGRAPH_INVMODE: invalid mode argument.
+ *         \c IGRAPH_EINVAL: invalid number of children.
+ *
+ * Time complexity: O(|V|+|E|), the
+ * number of vertices plus the number of edges in the graph.
+ * 
+ * \sa \ref igraph_tree() \ref igraph_lattice(), \ref igraph_star() for creating regular
+ * structures; \ref igraph_from_prufer() for creating arbitrary trees;
+ * \ref igraph_tree_game() for uniform random sampling of trees.
+ *
+ * \example examples/simple/igraph_symmetric_tree.c
+ */
+
+igraph_error_t igraph_symmetric_tree(igraph_t *graph, igraph_vector_int_t *branch_level,
+                igraph_tree_mode_t type) {
+    
+    igraph_vector_int_t edges = IGRAPH_VECTOR_NULL;
+    igraph_integer_t j, k;
+    igraph_integer_t idx = 0;
+    igraph_integer_t to = 1;
+    igraph_integer_t temp = 1;
+    igraph_integer_t vertices = 1;
+    igraph_integer_t current_vertex = 0;
+    igraph_integer_t vertex_nr = 0;
+
+    if (type != IGRAPH_TREE_OUT && type != IGRAPH_TREE_IN &&
+        type != IGRAPH_TREE_UNDIRECTED) {
+        IGRAPH_ERROR("Invalid mode argument", IGRAPH_EINVMODE);
+    }
+    if (!igraph_vector_int_empty(branch_level) && igraph_vector_int_min(branch_level) <= 0) {
+        IGRAPH_ERROR("Invalid number of children", IGRAPH_EINVAL);
+    }
+    
+    for(j = 0; j < igraph_vector_int_size(branch_level); ++j) {
```

Comment:
Let's also add a "TODO" comment to check for integer overflow. This will be done later once data structure size limits are properly defined for the library.